Modules/CTP/src/CTPTrendingTask.cxx

```diff
@@ -0,0 +1,223 @@
+
+  map<string, string> metadata; // can be empty
+
+  std::string run = std::to_string(t.activity.mId);
+  CTPRunManager::setCCDBHost("https://alice-ccdb.cern.ch");
+  mCTPconfig = CTPRunManager::getConfigFromCCDB(t.timestamp, run);
+
+  bool foundInputs = false;
+
+  for (auto& dataSource : mConfig.dataSources) {
+    auto mo = qcdb.retrieveMO(dataSource.path, dataSource.name, t.timestamp, t.activity);
+    if (!mo) {
+      ILOG(Info, Support) << "no MO object" << ENDM;
+      continue;
+    }
+    ILOG(Info, Support) << "Got MO " << mo << ENDM;
+    TObject* obj = mo ? mo->getObject() : nullptr;
+    if (obj) {
+      foundInputs = true;
+      mReductors[dataSource.name]->update(obj);
+    }
+  }
+
+  if (!foundInputs) {
+    ILOG(Info, Support) << "inputs not found" << ENDM;
+    return;
+  }
+
+  mTrend->Fill();
+}
+
+void CTPTrendingTask::generatePlots()
+{
+  if (mTrend == nullptr) {
+    ILOG(Info, Support) << "The trend object is not there, won't generate any plots." << ENDM;
+    return;
+  }
+
+  if (mTrend->GetEntries() < 1) {
+    ILOG(Info, Support) << "No entries in the trend so far, won't generate any plots." << ENDM;
+    return;
+  }
+
+  int index = 0;
+  int inputIndex = 0;
+  int classIndex = 0;
+
+  auto ctpClasses = mCTPconfig.getCTPClasses();
+  auto ctpInputs = mCTPconfig.getCTPInputs();
+
+  std::vector<int> nonExistingInputs = { 11, 12, 19, 20, 23, 24, 30, 31, 32, 33, 34, 35, 36, 45, 46, 47 }; // histograms for these inputs are not defined in the config
+
+  const int numberOfInputs = 48 - nonExistingInputs.size();
+  ILOG(Info, Support) << "numberOfInputs.  " << numberOfInputs << ENDM;
+
+  std::vector<int> trendedInputIndex;
+  std::vector<int> trendedClassIndex;
+  for (int i = 0; i < ctpInputs.size(); i++) {
+    auto input = ctpInputs[i];
+    if (std::find(nonExistingInputs.begin(), nonExistingInputs.end(), input.getIndex()) != std::end(nonExistingInputs)) {
+      ILOG(Info, Support) << "Input " << input.getIndex() << " should not exist!!" << ENDM;
+      continue;
+    }
+    if (trendedInputIndex.size() == 0)
+      trendedInputIndex.push_back(input.getIndex());
+    else if (std::find(trendedInputIndex.begin(), trendedInputIndex.end(), input.getIndex()) != std::end(trendedInputIndex))
+      continue;
+    else
+      trendedInputIndex.push_back(input.getIndex());
+  }
+
+  for (int i = 0; i < ctpClasses.size(); i++) {
+    auto ctpClass = ctpClasses[i];
+    if (i == 0)
+      trendedClassIndex.push_back(ctpClass.getIndex());
+    else if (std::find(trendedClassIndex.begin(), trendedClassIndex.end(), ctpClass.getIndex()) != std::end(trendedClassIndex))
+      continue;
+    else
+      trendedClassIndex.push_back(ctpClass.getIndex());
+  }
+
+  ILOG(Info, Support) << "Generating " << trendedClassIndex.size() + trendedInputIndex.size() << " plots." << ENDM;
+
+  for (const auto& plot : mConfig.plots) {
+
+    if (index < numberOfInputs && index != trendedInputIndex[inputIndex] - 1) {
```

Comment:
A bit of documentation on what is being done here would be nice. In particular given that there are two conditions leading in the end to the same duplicated code.